File: engine/operations/history.py
"""迁移历史：由 Python Engine 独占的 SQLite 状态持久化。"""
from __future__ import annotations

import secrets
from ..context import EngineContext
from ..storage.database import state_database as _database

def append(entry: dict, ports: EngineContext) -> str:
    history_id = "history_" + secrets.token_urlsafe(18)
    _database(ports).migration_history.append(history_id, entry)
    return history_id


def list_entries(ports: EngineContext) -> list[dict]:
    return _database(ports).migration_history.list_all()


def delete(history_id: str, ports: EngineContext) -> dict:
    return _database(ports).migration_history.delete(history_id)

File: crates/ferry-engine/src/operations/history.rs
//! 迁移历史：由 Engine 独占的 SQLite 状态持久化。
//!
//! 语义事实源：`engine/operations/history.py`。
//!
//! history_id = `history_` + `token_urlsafe(18)`（§2.2 第 14 条）。

use serde_json::Value;

use crate::operations::history_store::HistoryDeletion;
use crate::operations::plan_store::token_urlsafe;
use crate::operations::types::{EngineResult, Ports};
use crate::storage::database::state_database;

pub const HISTORY_ID_PREFIX: &str = "history_";

pub fn append(entry: &Value, ports: &Ports) -> EngineResult<String> {
    let history_id = format!("{HISTORY_ID_PREFIX}{}", token_urlsafe(18));
    state_database(ports.state_dir())?
        .migration_history
        .append(&history_id, entry)?;
    Ok(history_id)
}

pub fn list_entries(ports: &Ports) -> EngineResult<Vec<Value>> {
    state_database(ports.state_dir())?
        .migration_history
        .list_all()
}

pub fn delete(history_id: &str, ports: &Ports) -> EngineResult<HistoryDeletion> {
    state_database(ports.state_dir())?
        .migration_history
        .delete(history_id)
}
